Add lint for misstyped literal casting

mipli opened the pull request:
Adds a `mistyped_literal_suffixes` lint, to check if someone forgot to type `i` or `u` in a literal number suffix. For instance, wrote `3_32` instead of `3_u32`. 

Found no way good way to test if a literal number was negative, so the lint still suggest using `_u32` even if the number is negative. 

Solves https://github.com/rust-lang-nursery/rust-clippy/issues/3091

mipli commented on clippy_lints/src/literal_representation.rs:
```diff
@@ -211,11 +232,18 @@ impl<'a> DigitInfo<'a> {
             }
+            let suffix_hint = match self.suffix {
+                Some(suffix) if ["_8", "_32", "_64"].contains(&suffix) => {
+                    format!("_i{}` or `{}_u{}", &suffix[1..], hint, &suffix[1..])
```
Didn't know that rustfix used these suggestions like that. 

Do we want this lint to be automatically applied? With support for suggesting `_i128` instead of `_128` we can end up with false positives, since `_128` can be part of a properly formatted literal.

flip1995 replied:
Oh your right. `256_128` is a valid formatting for a int literal and would even be suggested by one of those lints for `256128`, which would then be linted to change it to `256_i128`. I have no idea how to handle this. We can't really know what to do with `_128` and I would suggest to just bail out and write this in the `Known Problems` section.

If you come up with a good idea, you can add it of course! If not just leave out the `_128` case.

mipli commented:
Thanks for the feedback, I'll get working on fixing those things!

flip1995 commented on clippy_lints/src/literal_representation.rs:
```diff
@@ -504,3 +551,11 @@ impl LiteralRepresentation {
 }
+
+fn is_mistyped_suffix(suffix: &str) -> bool {
+	["_8", "_16", "_32", "_64", "_128"].contains(&suffix)
```
NIT: indentation, please use 4 spaces.

mati865 replied:
I think calling `cargo fmt --all` should fix it

flip1995 replied:
Yeah probably, but please make sure to not add unrelated formatted files.

mipli replied:
I'll check the files I've touched to make sure they're all properly transformed. Unsure why it happened, my editor should have read the `.editorconfig` and used the settings there.

flip1995 commented on clippy_lints/src/literal_representation.rs:
```diff
@@ -136,17 +155,22 @@ impl<'a> DigitInfo<'a> {
+            if !float && (d == 'i' || d == 'u') ||
+                float && (d == 'f' || d == 'e' || d == 'E') ||
+                (!float && is_possible_suffix_index(&d, d_idx, len)) {
                 let suffix_start = if last_d == '_' { d_idx - 1 } else { d_idx };
```
Another interesting test case: `1__8`. The suggestion should be `1_i8`, but I think that it will be something different.

mipli replied:
I can probably add tests for `2__8` as well if we want to have lints for that as well. Don't think it will add too much complexity to this.

mipli replied:
> I think `is_possible_suffix_index` needs to check directly if the possible suffix is `"_8"`, `"_16"`, ... and not just if the length matches.

The check for those specific values are done using the `is_mistyped_suffix` function. So cases were the index function matches on e.g `_23` are discarded later on and the loop continues to check the next index.

flip1995 replied:
Yes please add a test for this. `2__8` gets currently linted (`consider: 28`). But should be `2_i8` in the future, while `2__9` should stay `consider: 29`

flip1995 replied:
> e.g _23 are discarded later on and the loop continues to check the next index.

Yes but I think that will fail on `__8`.

mipli replied:
It still returns `consider: 28` when you write `2__8`, so I think it does work as expected today. But that doesn't really matter now, since I'll be updating the handling of cases with double underscores and adding a few different test cases for those.

I'll add some extra tests for other strange uses of multiple underscores as well, just to document how we handle those.

flip1995 commented on clippy_lints/src/literal_representation.rs:
```diff
@@ -26,6 +26,25 @@ declare_clippy_lint! {
+/// **Why is this bad?** This is most probably a typo
+///
+/// **Known problems:**
+///		- Recommends unsigned suffix even if it's a negative number. Should only  recommend the signed suffix on those cases.
```
That is not true anymore. Maybe "always suggests singed integer", but I don't think this is really a problem.

mipli commented:
Fixed things mentioned above, and added test cases for some cases with strange use of underscores. 

Rewrote it a bit as well to make the code easier to read. Only issue I see with the rewrite is that we now do a `.split_at` for each character in the literal (up until the point where it finds an issue), rather than only when we suspect it might be an issue. Considering the speed of the `.split_at` function I don't think this is a problem, but it's worth being aware of the change.

flip1995 commented:
The suggestions are looking good now!

You could prevent the problem with calling the `split_at` method at every digit by combining your previous and current solution:
```rust
if ... ||
    !float && is_possible_suffix_index(..) {
...
fn is_possible_suffix_index(..) -> bool {
    d == "_" && <length-check> && is_misstyped_suffix(split_at());
}
```

Thanks to lazy evaluation the `split_at` method only gets called on the last 2-3 digits.

Also feel free to add as many commits to a PR as you like, you don't have to keep it at one. If necessary, they can be squashed before merging and it is easier for you and for the reviewer to see what changed since opening the PR.

flip1995 commented on clippy_lints/src/literal_representation.rs:
```diff
@@ -135,18 +155,25 @@ impl<'a> DigitInfo<'a> {
+            };
+            if !float && (d == 'i' || d == 'u') ||
+                float && (d == 'f' || d == 'e' || d == 'E') ||
+                !float && is_possible_suffix_index(&sans_prefix, suffix_start, len) {
```
`cargo clippy --all-targets --all-features -- -D clippy::all` fails on this:
```rust
if !float && A || !float && B
```
can be simplified to
```rust
if !float && (A || B)
```
https://travis-ci.org/rust-lang-nursery/rust-clippy/jobs/425293764#L1349-L1365

Everything else LGTM! I'm ready to merge this, after this is fixed.

Unrelated: I think the suggestion of the `nonminimal-bool` lint is wrong or at least way to complicated :thinking:

oli-obk replied:
We might be able to tune the minimization algorithm. I'll open an issue

mipli commented:
Is anything else required from me for this to be merged now? 

The AppVeyor build fails, but I'm not sure why it fails. On my local computer everything passes, so I can't properly debug that issue either.

oli-obk commented:
Appveyor is in flux right now, don't worry about it.
